src/front/spv/mod.rs

```diff
@@ -5322,7 +5338,11 @@ fn resolve_constant(
     gctx: crate::proc::GlobalCtx,
     constant: Handle<crate::Constant>,
 ) -> Option<u32> {
-    gctx.constants[constant].to_array_length()
+    match gctx.const_expressions[gctx.constants[constant].init] {
+        crate::Expression::Literal(crate::Literal::U32(id)) => Some(id),
+        crate::Expression::Literal(crate::Literal::I32(id)) => Some(id as u32),
+        _ => None,
+    }
```

Comment:
I've thought about this as well but decided against it since no evaluation would be necessary in the context this fn is used in according to the SPIR-V spec (they should all be 32-bit integer scalars > 0).